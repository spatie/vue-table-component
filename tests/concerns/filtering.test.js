import createVm from '../createVm';

describe('Filtering', () => {
    beforeEach(() => {
        window.localStorage.clear();

        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John', lastName: 'Lennon' },
                            { firstName: 'Paul', lastName: 'McCartney' },
                            { firstName: 'George', lastName: 'Harrison' },
                            { firstName: 'Ringo', lastName: 'Starr' }]"
                    sort-by="lastName"
                    sort-order="desc"
                >
                    <table-column show="firstName" label="First name"></table-column>
                    <table-column show="lastName" label="Last name" :filterable="false"></table-column>
                </table-component>
            </div>
        `;
    });

    it('can filter data', async () => {
        const table = await createVm(table => {
            table.filter = 'Paul';
        });

        expect(table.displayedRows).toHaveLength(1);
        expect(table.displayedRows[0].data.firstName).toBe('Paul');
    });

    it('can filter data in a case-insensitive way', async () => {
        const table = await createVm(table => {
            table.filter = 'paul';
        });

        expect(table.displayedRows).toHaveLength(1);
        expect(table.displayedRows[0].data.firstName).toBe('Paul');
    });

    it('will display a message if there are no matching rows', async () => {
        const table = await createVm(table => {
            table.filter = 'there are no rows that will match this';
        });

        expect(table.displayedRows).toHaveLength(0);
    });

    it('will not use columns that are not filterable', async () => {
        // Note: Only the firstName field is filterable
        // <table-column show="lastName" label="Last name" :filterable="false"></table-column>

        const table = await createVm(table => {
            table.filter = 'Lennon';
        });

        expect(table.displayedRows).toHaveLength(0);
    });

    it('can filter on another property', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John', songs: 72 },
                            { firstName: 'Paul', songs: 70 },
                            { firstName: 'George', songs: 22 },
                            { firstName: 'Ringo', songs: 2 }]"
                    sort-by="lastName"
                    sort-order="desc"
                >
                    <table-column show="firstName" label="First name" filter-on="songs"></table-column>
                </table-component>
            </div>
        `;

        const table = await createVm(table => {
            table.filter = '70';
        });

        expect(table.displayedRows).toHaveLength(1);
        expect(table.displayedRows[0].data.firstName).toBe('Paul');
    });

    it('can add a custom html class on the filter input', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John', songs: 72 },
                            { firstName: 'Paul', songs: 70 },
                            { firstName: 'George', songs: 22 },
                            { firstName: 'Ringo', songs: 2 }]"
                    sort-by="lastName"
                    sort-order="desc"
                    filter-input-class="my-filter-class"
                >
                    <table-column show="firstName" label="First name" filter-on="songs"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('hides the filter when no columns are filterable', async () => {
        document.body.innerHTML = `
            <div id="app">
                <table-component
                    :data="[{ firstName: 'John', songs: 72 },
                            { firstName: 'Paul', songs: 70 },
                            { firstName: 'George', songs: 22 },
                            { firstName: 'Ringo', songs: 2 }]"
                    sort-by="lastName"
                    sort-order="desc"
                >
                    <table-column show="firstName" label="First name" :filterable="false"></table-column>
                </table-component>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});
