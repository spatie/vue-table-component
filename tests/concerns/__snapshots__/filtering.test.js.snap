// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtering can add a custom html class on the filter input 1`] = `

<div id="app">
  <div class="table-component">
    <div class="table-component__filter">
      <input type="text"
             placeholder="Filter tableâ€¦"
             class="table-component__filter__field my-filter-class"
      >
      <!---->
    </div>
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table sorted by lastName (descending)
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
          <tr>
            <td>
              George
            </td>
          </tr>
          <tr>
            <td>
              Ringo
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
    <!---->
  </div>
</div>

`;

exports[`Filtering hides the filter when no columns are filterable 1`] = `

<div id="app">
  <div class="table-component">
    <!---->
    <div class="table-component__table-wrapper">
      <table class="table-component__table">
        <caption role="alert"
                 aria-live="polite"
                 class="table-component__table__caption"
        >
          Table sorted by lastName (descending)
        </caption>
        <thead class="table-component__table__head">
          <tr>
            <th role="columnheader"
                aria-sort="none"
                class="table-component__th table-component__th--sort"
            >
              First name
            </th>
          </tr>
        </thead>
        <tbody class="table-component__table__body">
          <tr>
            <td>
              John
            </td>
          </tr>
          <tr>
            <td>
              Paul
            </td>
          </tr>
          <tr>
            <td>
              George
            </td>
          </tr>
          <tr>
            <td>
              Ringo
            </td>
          </tr>
        </tbody>
        <tfoot>
        </tfoot>
      </table>
    </div>
    <!---->
    <div style="display: none;">
      <!---->
    </div>
    <!---->
  </div>
</div>

`;
